### DISCUSSION TITLE
Default file system files

### DISCUSSION THREAD

**Discussion by bensuffolk** [HASwitchPlate/openHASP#346] [Ideas]
I know that on a Factory Reset, or the first time the firmware is uploaded onto a clean device some files are created, and I could define the text for these files, for instance with HASP_PAGES_JSONL

But it would be really nice to be able to put some files into a custom build directory (including graphics) that got put in the filesystem instead of the default files being created.

This way when a factory reset it done, a complete set of files would be there ready to go. This would mean that you did not have to go to the plate and upload the files to it.

**Comment by fvanroie** [comment]
There are currently `HASP_ONLINE_CMD`, `HASP_OFFLINE_CMD` and `HASP_PAGES_JSONL` defined in the code.
Provisioning a plate through the Factory Reset command hasn't crossed my mind, as I think both tasks serve a different purpose.

Adding text files to the filesystem is easy since these filenames are known and the size of the data is small.
Adding binary files isn't as straightforward, as you'd need pre-compiler flags like:
```yaml
board_build.embed_files =
    data/edit.htm.gz
    data/style.css.gz
    data/script.js.gz
    data/cert/x509_crt_bundle.bin
```
The size gets tagged onto the firmware binary which is limited to 1664KB-2MB depending on the flash partitioning scheme.
So depending on the size of the provisioned files, you might need to adjust the partition size too.

Some suggestions for provisioning files:

- If size isn't an issue, I could imaging you could put a similar **uncompressed** `.zip` file in the `/data/` folder and include it like the `.gz` files. Then upon filesystem initialization, `unzip` that byte-array to the flash partition much like `filesystemUnzip` does but from the flash byte-array instead...

- You can also upload the files as **uncompressed** `default.zip` and then run `unzip default.zip` command to extract the files.

- There is the option for using the FTP-server #283 to easily transfer files to and from the plate.

- There is also the Upload Filesystem option in the Firmware Upgrade page, but I really haven't tested that thoroughly. It requires creating a LittleFS partition file using `mklittlefs` and uploading that file.
![image](https://user-images.githubusercontent.com/15969459/174745999-0b4ad6cf-f019-4bc2-b0a1-5631a33156bb.png)

**Reply by bensuffolk** [reply]
Some interesting options there. I was thinking along the lines of the default.zip file as I have 2 graphics files I want to add in, so the defines are not quite what I need.

I'm happy to save firmware space by sacrificing some unused (for me) functionality like FTP so I think thats the best option for me to look into.

I will check out the LittleFS partition file (I had not noticed that option during upload), as if that happens to be smaller than an uncompressed zip file, it might be a better option to include that in the firmware bundle instead.

**Reply by bensuffolk** [reply]
Just a thought, a more interesting option would be to have file upload data via MQTT. That would be pretty cool. 

So something like /plates/xyz/fileupload/filename then the file contents as the payload.

**Reply by fvanroie** [reply]
MQTT has a limited receive buffer of a few KB which makes transferring files a challenge. I've seen it done, it's just not as straightforward...
Why would you dismiss the FTP option? That's literally the File Transfer Protocol :)

**Reply by bensuffolk** [reply]
I'm writing a system that uses a microcontroller as a master device talking to the wall plates. So I have written a simple MQTT broker for it and I guess I could use an FTP client on that as well. I was just trying to minimise the code on it, but you're probably right. It would end up being the right way to do it.
